stablehlo/dialect/TypeInference.cpp

```diff
@@ -1069,186 +1069,62 @@ LogicalResult inferOutfeedOp(MLIRContext* context, Optional<Location> location,
-  SmallVector<TensorType> initValueTypes{llvm::map_range(
-      initValues.getTypes(),
-      [](Type t) -> TensorType { return t.cast<TensorType>(); })};
-  uint64_t numInputs = inputs.size();
-
-  // Check for unranked tensors in input operands.
-  int64_t rankedInputIdx = -1;
-  for (uint64_t inputIdx = 0; inputIdx < numInputs; ++inputIdx) {
-    if (inputArgTypes[inputIdx].hasRank()) {
-      rankedInputIdx = inputIdx;
-      break;
-    }
-  }
-
-  bool allInputsUnranked = (rankedInputIdx == -1);
-
-  // P1.
-  if (!allInputsUnranked) {
-    for (uint64_t inputIdx = 0; inputIdx < numInputs; ++inputIdx) {
-      if (failed(mlir::verifyCompatibleShape(inputArgTypes[rankedInputIdx],
-                                             inputArgTypes[inputIdx]))) {
-        return emitOptionalError(
-            location, "expects all inputs to have compatible shapes. Shape at",
-            " input-index ", inputIdx,
-            " is not compatible with shape at input-index ", rankedInputIdx);
-      }
-    }
-  }
-
-  // P2.
-  DenseSet<int64_t> dimensionsToReduceSet;
-  for (int64_t dimension : dimensions.getValues<int64_t>()) {
-    if ((!allInputsUnranked &&
-         dimension >= inputArgTypes[rankedInputIdx].getRank()) ||
-        dimension < 0) {
-      return emitOptionalError(
-          location, "Out-of-bounds dimension ", dimension,
-          " for input-tensor rank: ", inputArgTypes[rankedInputIdx].getRank());
-    }
-
-    if (!dimensionsToReduceSet.insert(dimension).second) {
-      return emitOptionalError(location,
-                               "Duplicate reduction dimension: ", dimension);
-    }
-  }
-
-  // P3.
-  SmallVector<int64_t> newDimensions;
-  if (!allInputsUnranked) {
-    for (int inputIdx = 0; inputIdx < inputArgTypes[rankedInputIdx].getRank();
-         ++inputIdx) {
-      if (!dimensionsToReduceSet.count(inputIdx)) {
-        newDimensions.push_back(
-            inputArgTypes[rankedInputIdx].getDimSize(inputIdx));
-      }
+  for (auto input : inputs) {
+    Type elementTy = getElementTypeOrSelf(input.getType());
+    auto rankedTy = input.getType().dyn_cast<RankedTensorType>();
+    if (!rankedTy) {
+      inferredReturnShapes.emplace_back(elementTy);
+      continue;
     }
-  }
-
-  Block& block = body.front();
-  if (failed(verifyReducerShape(location, block, inputArgTypes, initValueTypes,
-                                numInputs, newDimensions, allInputsUnranked)))
-    return failure();
 
-  for (uint64_t inputIdx = 0; inputIdx < numInputs; ++inputIdx) {
-    TensorType inputType = inputArgTypes[inputIdx];
-    Type elementType = inputType.getElementType();
-    if (inputType.hasRank())
-      inferredReturnShapes.emplace_back(newDimensions, elementType);
-    else
-      inferredReturnShapes.emplace_back(elementType);
+    int64_t rank = rankedTy.getRank();
+    llvm::SmallVector<bool, 4> dimsMask(rank, false);
+    for (int64_t dim : dimensions.getValues<int64_t>()) dimsMask[dim] = true;
```

Comment:
Do we verify dimensions attribute before reaching here to make sure that this doesn't lead to crash because of invalid arguments? (I haven't read this change thoroughly)

Reply:
Great question.
The current commit was (for `Reduce` and `ReduceWindow`): 1. put all verifications in verifier. 2. infer shape assuming everything is valid in this shape function.
After a second thought now, I think this is probably not OK when the shape functions is called alone (for example in the `create<ReduceOp>` case which this PR is trying to address. Thus, I have no choice but to do the necessary validation again here: put them in an util function and called both by verifier and shape function). 
Will update in a next commit, then make this comment as resolved.

Reply:
Updated.
ReduceOp: introduce new util `verifyReduceOpInputsAndInferShape()`
ReduceWindow: introduce new util `verifyReduceWindowOpInputsAndInferWindow()`
In each op, both the verifier & shape functions calls this util function.